pkg/acpi/coreboot/madt.go

```diff
@@ -0,0 +1,169 @@
+)
+const (
+	entryLAPIC         madtType = 0
+	entryIOAPIC        madtType = 1
+	entryISOR          madtType = 2 // pronounced: eyesore, for reasons.
+	entryBOGUS         madtType = 3
+	entryNMI           madtType = 4
+	entryLAPICOverRide madtType = 5
+	entryLX2APIC       madtType = 9
+	entryLX2NMI        madtType = 10
+
+	sizeEntryLAPIC         uint8 = 8
+	sizeEntryIOAPIC        uint8 = 12
+	sizeEntryISOR          uint8 = 10 // pronounced: eyesore, for reasons.
+	sizeEntryNMI           uint8 = 6
+	sizeEntryLAPICOverRide uint8 = 12
+	sizeEntryLX2APIC       uint8 = 16
+	sizeEntryLX2NMI        uint8 = 12
+)
+
+var (
+	// The MADT signature is APIC.
+	// Another failure of vision; they never thought there'd
+	// be more than one APIC I guess.
+	sigMADT = "APIC"
+	info    = map[madtType]struct {
+		name string
+		size uint8
+	}{
+		entryLAPIC:         {"LAPIC", sizeEntryLAPIC},
+		entryIOAPIC:        {"IOAPIC", sizeEntryIOAPIC},
+		entryISOR:          {"ISOR", sizeEntryISOR},
+		entryNMI:           {"NMI", sizeEntryNMI},
+		entryLAPICOverRide: {"LAPICOverRide", sizeEntryLAPICOverRide},
+		entryLX2APIC:       {"LX2APIC", sizeEntryLX2APIC},
+		entryLX2NMI:        {"X2NMI", sizeEntryLX2NMI},
+	}
+)
+
+type subtable struct {
+	t madtType
+	b []byte
+}
+
+// MADT defines the madt struct.
+type MADT struct {
+	acpi.Table
+	// base is the base address of the MADT struct in physical memory.
+	base      int64
+	subtables []subtable
+}
+
+var _ Corebooter = &MADT{}
+
+func init() {
+	corebooters["APIC"] = NewMADT
+}
+
+// rv reverses a slice
+func rv(b []byte) []byte {
+	for i := 0; i < len(b)/2; i++ {
+		j := len(b) - i - 1
+		b[i], b[j] = b[j], b[i]
+	}
+	return b
+}
+
+// NewMADT returns a MADT, from a Table, iff the signature is correct.
+func NewMADT(t acpi.Table) (Corebooter, error) {
+	if t.Sig() != sigMADT {
+		return nil, fmt.Errorf("%s is not a MADT", t.Sig())
+	}
+	b := t.TableData()
+	Debug("NEWMADT Subtables: %d bytes", len(b))
+	if len(b) < int(minMADT) {
+		return nil, fmt.Errorf("Table is only %d bytes, need %d", len(t.TableData()), minMADT)
+	}
+	b = b[8:]
+	Debug("NEWMADT Subtables: after removing 8 bytes of fluff, %d bytes", len(b))
+	m := &MADT{Table: t, base: t.Address()}
+	for len(b) > 0 {
+		Debug("NewMADT loop: %d bytes\n", len(b))
+		if len(b) < 2 {
+			return nil, fmt.Errorf("subtable is too short: have %d bytes, need 2", len(b))
+		}
+		l := b[1]
+		typ := madtType(b[0])
+		Debug("subtable: %d bytes, l is %#x, t is %#x", len(b), l, t)
+		if int(typ) > len(info) || typ == entryBOGUS {
+			Debug("Bad type: %#x, offset %#x, skipping %d bytes", typ, len(t.TableData())-len(b), l)
+			b = b[l:]
+			continue
+		}
+		if l != info[typ].size {
+			return nil, fmt.Errorf("Type %s has size %d should be %d", info[typ].name, l, info[typ].size)
+		}
+		if len(b) < int(l) {
+			return nil, fmt.Errorf("Type %s has buf length %d should be %d", info[typ].name, len(b), info[typ].size)
+		}
+		m.subtables = append(m.subtables, subtable{t: typ, b: b})
```

Comment:
for this: subtable{t: typ, b: b}

probably doesn't matter, but does that create a new byte slice consisting of b's current contents, including all of the entries after the one you are handling?  since you know the size of the entry (info[typ].size), it probably doesn't matter.  but if you did len(sub_t.b) you might get more than info[sub_t.type].size.

Reply:
oops, yeah, should be b[:l]